src/primitives/gf32.rs

```diff
@@ -285,58 +256,43 @@ impl AsRef<u8> for Fe32 {
     fn as_ref(&self) -> &u8 { &self.0 }
 }
 
-/// Implements $op for the 2x2 matrix of type by ref to type
-macro_rules! impl_op_matrix {
-    ($op:ident, $op_fn:ident, $call_fn:ident) => {
-        impl ops::$op<Fe32> for Fe32 {
-            type Output = Fe32;
-            #[inline]
-            fn $op_fn(self, other: Fe32) -> Fe32 { self.$call_fn(other) }
-        }
-
-        impl ops::$op<Fe32> for &Fe32 {
-            type Output = Fe32;
-            #[inline]
-            fn $op_fn(self, other: Fe32) -> Fe32 { self.$call_fn(other) }
-        }
-
-        impl ops::$op<&Fe32> for Fe32 {
-            type Output = Fe32;
-            #[inline]
-            fn $op_fn(self, other: &Fe32) -> Fe32 { self.$call_fn(*other) }
+impl super::Field for Fe32 {
+    const ZERO: Self = Fe32::Q;
+    const ONE: Self = Fe32::P;
+    const GENERATOR: Self = Fe32::Z;
+    const MULTIPLICATIVE_ORDER: usize = 31;
+    const MULTIPLICATIVE_ORDER_FACTORS: &'static [usize] = &[1, 31];
+
+    fn _add(&self, other: &Fe32) -> Fe32 { Fe32(self.0 ^ other.0) }
+    // Subtraction is the same as addition in a characteristic-2 field
+    fn _sub(&self, other: &Fe32) -> Fe32 { Fe32(self.0 ^ other.0) }
+    fn _mul(&self, other: &Fe32) -> Fe32 {
+        if self.0 == 0 || other.0 == 0 {
+            Fe32(0)
+        } else {
+            let log1 = LOG[self.0 as usize];
+            let log2 = LOG[other.0 as usize];
+            Fe32(LOG_INV[((log1 + log2) % 31) as usize])
```

Comment:
The bare `31` here and in `_div` feels like magic. Should these be named constants?

Reply:
Sure, I can use `LOG_ENV.len()`. But 31 is the number of nonzero elements in a 32-element field. I think it's ok to use both 31 and 32 as "magic numbers" within this crate.

Reply:
Ah, we have the `Self::MULTIPLICATIVE_ORDER` constant available.